server/server.go

```diff
@@ -61,12 +62,40 @@ func requestHandler(router proxy.Router, logger *apexlog.Logger) func(http.Respo
 			}
 		}
 
+		var reader io.ReadCloser
+		var writer http.ResponseWriter
+		if reqres.AddCompressionType != util.CompressionTypeNone {
+			if reqres.Response.Header.Get(headerContentEncodingKey) == "gzip" {
+				reader, err = util.NewGzipDecodingReader(reqres.Response.Body)
```

Comment:
If the client is requesting gzip and the server is sending gzip, are we needlessly decoding and recompressing the body here? If yes, I think a shortcut should be added for this case.

Reply:
…and also for the similar brotli requested + brotli served from origin case

Reply:
> If the client is requesting gzip and the server is sending gzip, are we needlessly decoding and recompressing the body here? If yes, I think a shortcut should be added for this case.

No, it's handled here and no additional compression is done:

```go
	case acceptsGzip:
		switch contentEncoding {
		case "gzip":
			return CompressionTypeNone
```

Reply:
Actually, I have some doubts about that when the user agent is accepting both Brotli and gzip. Input for the quoted switch-case comes from the new function `acceptsEncodingFromString` that seems to return `acceptsBrotli`, if the header value contains both `br` and `gzip`, which seems to lead to re-compressing gzip with Brotli. Could you double check that, and maybe write another test case?

Reply:
We want to recompress gzip with brotli in the case the client accepts both.

Reply:
Or did you mean that the brotli compression is erroneously applied to the gzipped body, not the decompressed plaintext?

Reply:
@markonen OK, good to have that clarified. I apparently had a different idea about wanting/accepting, and the logic seems to be correct if we want to always use Brotli when possible. No, we are not missing decompression.

Reply:
That won't happen, but yeah this not clear here. I think a .RemoveCompression on the requestResponse would make this clearer, since both decisions to add and remove would be explicitly done in the single function.

Reply:
Changed in 2174624.